### PR TITLE
a simple background identifier

### PR DESCRIPTION
**Pull request by maxtaco**
- has an Add(uid) and Remove(uid) interface
- is constantly reidentifying users, faster if there are soft errors,
  and slightly faster if there are hard errors.
- maintains a heap-ordered work queue
- a pretty rigorous test to see that it works as adverised (most of the work :( )

### CODE REVIEW

**Comment by keybase-ci-visdiff** [comment]
The commits d673606df51e...a7502672816a introduce visual changes on linux. <details><summary>:mag_right: 1 changed</summary>

 * Changed: **Chat_Input-Emoji_Open** [(before)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1481347594989-d673606df51e-a7502672816a-linux/Chat_Input-Emoji_Open-d673606df51e.png) [(after)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1481347594989-d673606df51e-a7502672816a-linux/Chat_Input-Emoji_Open-a7502672816a.png)  
   Chat_Input-Emoji_Open comparison
</summary>

**Comment by keybase-ci-visdiff** [comment]
The commits d673606df51e...6f8a33b4903c introduce visual changes on linux. <details><summary>:mag_right: 1 changed</summary>

 * Changed: **Chat_Input-Emoji_Open** [(before)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1481373328733-d673606df51e-6f8a33b4903c-linux/Chat_Input-Emoji_Open-d673606df51e.png) [(after)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1481373328733-d673606df51e-6f8a33b4903c-linux/Chat_Input-Emoji_Open-6f8a33b4903c.png)  
   Chat_Input-Emoji_Open comparison
</summary>

File: go/engine/bg_identifier.go [hunk 1]
```diff
@@ -0,0 +1,299 @@
...
+)
+
+type bgiUser struct {
+	uid     keybase1.UID
+	nextRun time.Time
+	// The index is needed by update and is maintained by the heap.Interface methods.
+	index int // The index of the item in the heap.
+}
+
+// A PriorityQueue implements heap.Interface and holds Items.
```
Comment (patrickxb) [inline go/engine/bg_identifier.go]: holds *bgiUser

File: go/engine/identify2_with_uid.go [hunk 2]
```diff
@@ -280,7 +281,13 @@ func (e *Identify2WithUID) Run(ctx *Context) (err error) {
 func (e *Identify2WithUID) run(ctx *Context) {
 	err := e.runReturnError(ctx)
 	e.unblock( /* isFinal */ true, err)
-	ctx.IdentifyUI.Cancel() // always cancel IdentifyUI to allow clients to clean up
+
+	// always cancel IdentifyUI to allow clients to clean up.
+	// If no identify was specified (because running the background)
```
Comment (patrickxb) [inline go/engine/identify2_with_uid.go]: If no IdentifyUI

File: go/engine/bg_identifier.go [hunk 3]
```diff
@@ -0,0 +1,299 @@
...
+type priorityQueue []*bgiUser
+
+var _ heap.Interface = (*priorityQueue)(nil)
+
+func (pq priorityQueue) Len() int { return len(pq) }
+
+func (pq priorityQueue) Less(i, j int) bool {
+	// We want Pop to give us the not lowest timeout, so user less than
```
Comment (patrickxb) [inline go/engine/bg_identifier.go]: this comment a little confusing...

**Comment by keybase-ci-visdiff** [comment]
The commits d673606df51e...13637b558966 introduce visual changes on linux. <details><summary>:mag_right: 1 changed</summary>

 * Changed: **Chat_Input-Emoji_Open** [(before)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1481419641271-d673606df51e-13637b558966-linux/Chat_Input-Emoji_Open-d673606df51e.png) [(after)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1481419641271-d673606df51e-13637b558966-linux/Chat_Input-Emoji_Open-13637b558966.png)  
   Chat_Input-Emoji_Open comparison
</summary>

**Comment by keybase-ci-visdiff** [comment]
The commits d673606df51e...6ff4d7f08def introduce visual changes on linux. <details><summary>:mag_right: 1 changed</summary>

 * Changed: **Chat_Input-Emoji_Open** [(before)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1481421027280-d673606df51e-6ff4d7f08def-linux/Chat_Input-Emoji_Open-d673606df51e.png) [(after)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1481421027280-d673606df51e-6ff4d7f08def-linux/Chat_Input-Emoji_Open-6ff4d7f08def.png)  
   Chat_Input-Emoji_Open comparison
</summary>

**Comment by keybase-ci-visdiff** [comment]
The commits d673606df51e...6ff4d7f08def introduce visual changes on linux. <details><summary>:mag_right: 1 changed</summary>

 * Changed: **Chat_Input-Emoji_Open** [(before)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1481422098392-d673606df51e-6ff4d7f08def-linux/Chat_Input-Emoji_Open-d673606df51e.png) [(after)](https://keybase-jenkins-visdiff.s3.amazonaws.com/1481422098392-d673606df51e-6ff4d7f08def-linux/Chat_Input-Emoji_Open-6ff4d7f08def.png)  
   Chat_Input-Emoji_Open comparison
</summary>

**Pull request merged**